### ISSUE TITLE
remove support for 'startSessionCommand' contribution

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode#33791]
With the introduction of the `resolveDebugConfiguration` method on type `DebugConfigurationProvider` in the debug API, it is no longer necessary to contribute a `startSessionCommand` on the `debuggers` contribution point. Therefore we are deprecating the `startSessionCommand` in this milestone and plan to drop it in October 2017.

**Comment by isidorn** [comment]
ping @roblourens since we have now removed support for the startSessionCommand so it would be great if chrome adopts this change and starts using DebugConfigurationProviders. There are great examples in the Mock Debug in Node Debug

**Comment by DonJayamanne** [comment]
The ```startSessionCommand``` command provided the ability to debug even without having to open a workspace.
I have now adopted the ```DebugCOnfigurationProvider``` and registered it, and when testing debugging of a simple python file (in the [python extension](https://github.com/DonJayamanne/pythonVSCode)) I don't get what I expect.

The ```resolveDebugConfigration``` method is invoked and the ```provideDebugConfigurations``` is not.
Also, the ```debugConfiguration``` object passed into the ```resolveDebugCOnfiguration``` method is an empty object.

Is the solution for me to check if it's an empty object and then inject the other properties that would be required by my debugger?

Or am I missing something in my workflow/implementation?

**Comment by weinand** [comment]
@DonJayamanne yes, checking for the empty object is a correct approach. It has **not** changed from the `startSessionCommand` approach. 

`provideDebugConfigurations` is only called if a new launch.json needs to be created (which is not the case when you just want to debug a python file from the editor with F5). This behaviour has not changed with the introduction of the `DebugCOnfigurationProvider`.

See what we do in node-debug: https://github.com/Microsoft/vscode-node-debug/blob/ae3618de23acc4ce6e81a5d65e56ed76f9ef69da/src/node/extension/configurationProvider.ts#L34

**Comment by isidorn** [comment]
fyi @daviwil @rkeithhill

**Comment by gregg-miskelly** [comment]
@isidorn the new debugging API was introduced in the August release (that shipped in early September) and the old API will be removed in the October release (which ships in early November). Is that correct?

**Comment by isidorn** [comment]
@gregg-miskelly correct

**Comment by DanTup** [comment]
@isidorn Thanks; I'm hoping to work on this this weekend - hoped to have done it already!

Are the APIs already removed in insiders?

**Comment by isidorn** [comment]
@DanTup yes, in insiders the command support is already removed

**Comment by testforstephen** [comment]
In the latest vscode-insider (2017-10-11), if no launch.json file exists in current workspace, click `F5` and it seems nothing happens. Is there any possible to pop up the `Add Configuration` dropdown menu to guide user?

Another issue is that when a previous debug session is still on starting, click `F5` again will cause duplicated debug sessions for the same debug configuration. 

For example, the first activation of the java debug server may cost several seconds to start up. During this period, if user clicks `F5` for several times, VSCode will launch duplicated debug sessions for the same debug configuration. 

Can VSCode client have the possible to fix it? or passing a configuration id for the selected debug config in the API `resolveDebugConfiguration` so that the debugger can check duplicate by itself?

**Comment by testforstephen** [comment]
@isidorn sure. open two new issues for this.
https://github.com/Microsoft/vscode/issues/36045
https://github.com/Microsoft/vscode/issues/36044

**Comment by DanTup** [comment]
I think I've managed to move over from this, but when converting from `startSessionCommand` (which used to take a `FlutterLaunchRequestArguments`, a sub-class of `DebugProtocol.LaunchRequestArguments`) I had to put this nasty cast in so I can read/write the launch arguments:

```ts
resolveDebugConfiguration(folder: WorkspaceFolder | undefined, debugConfig: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {
	this.setupDebugConfig(<FlutterLaunchRequestArguments><any>debugConfig);
	// ...
```

The only other examples I found using this all just index into the `DebugConfiguration` but I really want to stick with my nice types. Is this cast a good way to do it, or have I done something bad? (It does appear to work).

**Comment by isidorn** [comment]
Shouldnt `FlutterLaunchRequestArguments` extend from `DebugConfiguration` at the end?
Not sure how to make the types flow nicely, but at the end `DebugConfiguration` is just the required part of the configuraiton that the vscode is aware of. Debug adapters are aware of additional fields on top of the basic ones.

**Comment by DanTup** [comment]
I did try that, but then I just errors at the other end because my DebugSession's `args` property was no longer a `LaunchRequestArgs`:

```ts
export class DartDebugSession extends DebugSession {
```

I'ts not obvious to me if there should be a relationship between `DebugConfig` and `DebugProtocol.LaunchRequestArguments`.. one is being built at the Code side and the other being passed to the debug adapter, but in code they're totally unrelated.

I'll leave my cast in for now since it seems to work. Looks like other adapters might just indexing into the object with string keys so they don't see this (but I really don't want to do that, I already have a nice class!).

**Comment by isidorn** [comment]
fyi @DustinCampbell 

@rkeithhill I have only pinged @gregg-miskelly on this issue, forgot Dustin. Thanks for pointing out

**Comment by gregg-miskelly** [comment]
Yup, we have this tracked with https://github.com/OmniSharp/omnisharp-vscode/issues/1787

**Comment by rkeithhill** [comment]
OK, got the PowerShell extension updated to work with these changes - PR coming shortly

**Comment by DanTup** [comment]
Dart Code is now all updated (as of v2.4.3, published yesterday) 🙂

**Comment by isidorn** [comment]
fyi @pieandcakes

**Comment by pieandcakes** [comment]
@isidorn I don't think we implement this. Thanks for the heads up!

**Comment by rkeithhill** [comment]
BTW it seems we have to set our extension's min vscode engine version to 1.17.  In reading release notes, it would appear this was intro'd in 1.16 but when I peruse the vscode.d.ts file for the v1.16.0 tag, there is no `vscode.debug`.  However there  is in `vscode.proposed.d.ts`.

This makes the compat cross-over a bit tight.  I mean 1.17 came out less than a month ago but when 1.18 drops we have no choice but to update to a min version of 1.17.  That leaves folks who updated VSCode less than a month ago, not able to debug with the latest PowerShell extension (until they update VSCode).

**Comment by weinand** [comment]
@rkeithhill yes, I share your concern and in the future we will always use a 2 month grace period.
However, we will not restore the already removed functionality (like 'startSessionCommand').

@isidorn lets make sure that we remove all future deprecations only after a 2 month grace period.

**Comment by ramya-rao-a** [comment]
@rkeithhill 

> This makes the compat cross-over a bit tight. I mean 1.17 came out less than a month ago but when 1.18 drops we have no choice but to update to a min version of 1.17. That leaves folks who updated VSCode less than a month ago, not able to debug with the latest PowerShell extension (until they update VSCode).

As far as I understood extension updates, an update with engine version set to say 1.17 will not even show up as an update to someone using VS Code 1.16. They will never get the extension update until they update VS Code. So they should be able to debug just fine.
@sandy081 can confirm this.

**Comment by sandy081** [comment]
Yes, we only install the compatible extensions for the given VS Code version by checking their `engine` property.
